### PR TITLE
addition of: list-commits and show-commit commands

### PR DESCRIPTION
**Pull request by nammn**
## To change/add
- [x] sort list-commits desc
- [x] time format in human readable format (4h ago...)
- [x] no commits feedback

#### adds:
- `commits`
- `show-commit <number>`
commands

#### changes:
- renames feature flag from generations to branches

#### preparation
- bootstrapped controller
```
export JUJU_DEV_FEATURE_FLAGS=generations
juju deploy ubuntu
juju add-branch bla
juju add-branch blub
juju track bla ubuntu
juju config ubuntu hostname=random
juju commit bla
```

#### list-commits test
```
❯ juju commits
Commit	Committed at              	Committed by	Branch name
3     	07 Nov 2019 14:41:41+01:00	admin       	bla    
```
should not show anything other than bla, because blub did not get committed

It should not show aborted branches as well!

no branches committed yet
```
~/golang/src/github.com/juju/juju list-commit
❯ juju commits                      
Commit	Committed at	Committed by	Branch name
```
--> might want to change that to a proper message

#### show-commits test
```
> juju show-commit 3

branch:
  bla:
    applications:
    - application: ubuntu
      units:
        tracking:
        - redis/0
      config:
        hostname: bla
committed-at: 08 Nov 2019 10:14:10+01:00
committed-by: admin
created: 08 Nov 2019 10:13:04+01:00
created-by: admin
```

err case:

```
## Documentation change
- rename feature flag from generations to branches

❯ juju show-commit 10 
ERROR generation_id 10 in model "default" not found
```

### CODE REVIEW

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 1]
```diff
@@ -276,6 +277,74 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
...
+		return generationCommitResultError(err)
+	}
+
+	branch, err := api.model.CommittedBranch(arg.GenerationId)
+	if err != nil {
+		result.Error = common.ServerError(err)
+		return result, nil
+	}
+
+	//TODO: convert them
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: TODO: Relevant?

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 2]
```diff
@@ -276,6 +277,74 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
...
+	_, err = api.oneBranchInfo(branch, true)
+	if err != nil {
+		return generationCommitResultError(err)
+	}
+
+	//result.GenerationCommit. = generation.Applications
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: Remove comment

File: cmd/juju/model/listcommits.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/juju/juju/core/model"
+	"github.com/juju/juju/juju/osenv"
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
```
Comment (SimonRichardson) [inline cmd/juju/model/listcommits.go]: You need to format these correctly.

File: cmd/juju/model/listcommits.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+)
+
+const (
+	listCommitsSummary = "Lists commits history"
+	listCommitsDoc     = `
+commits shows the timeline of changes to the model that occurred through branching.
```
Comment (SimonRichardson) [inline cmd/juju/model/listcommits.go]: Start with a capital letter.

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 5]
```diff
@@ -276,6 +277,74 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
 	return result, nil
 }
 
+// ShowCommit will return details a commit given by its generationId
+// An error is returned if no commit can be found corresponding to a branch.
+func (api *API) ShowCommit(arg params.GenerationId) (params.GenerationCommitResult, error) {
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: Do you need to bump the API, I believe you would need to, but you should check.

File: cmd/juju/model/showcommit.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/juju/juju/core/model"
+	"time"
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	showCommitsSummary = "Shows the commit content"
```
Comment (SimonRichardson) [inline cmd/juju/model/showcommit.go]: Of what?

**Comment by nammn** [comment]
!!build!!

windows failed...

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 7]
```diff
@@ -276,6 +289,72 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
...
+	}
+
+	result.GenerationCommit = generationCommit
+
+	return result, nil
+}
+
+// ListCommits will return the commits, hence only branches with generation_id higher than 0
+func (api *APIV3) ListCommits() (params.GenerationCommitResults, error) {
+	result := params.GenerationCommitResults{}
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: tiny nit:

```suggestion
	var result params.GenerationCommitResults
```

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 8]
```diff
@@ -324,6 +403,23 @@ func (api *API) oneBranchInfo(branch Generation, detailed bool) (params.Generati
 	}, nil
 }
 
+func (api *APIV3) getGenerationCommit(branch Generation) (params.GenerationCommit, error) {
+
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: nit: remove empty line.

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 9]
```diff
@@ -276,6 +289,72 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
 	return result, nil
 }
 
+// ShowCommit will return details a commit given by its generationId
+// An error is returned if either no branch can be found corresponding to the generation id.
+// Or the generation id given is below 1.
+func (api *APIV3) ShowCommit(arg params.GenerationId) (params.GenerationCommitResult, error) {
```
Comment (SimonRichardson) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]:
```suggestion
func (api *API) ShowCommit(arg params.GenerationId) (params.GenerationCommitResult, error) {
```

API is actually APIV4 and you end up masking it.

File: cmd/juju/model/listcommits.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+// Init implements part of the cmd.Command interface.
+func (c *CommitsCommand) Init(args []string) error {
+	lArgs := len(args)
+	if lArgs > 0 {
+		return errors.Errorf("expected no arguments, but got %v", lArgs)
+	}
+
+	// If use of ISO time not specified on command line, check env var.
+	if !c.isoTime {
+		var err error
```
Comment (SimonRichardson) [inline cmd/juju/model/listcommits.go]: Move err to inside the if statement below.

File: cmd/juju/model/listcommits.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+	table.AddRow("Commit", "Committed at", "Committed by", "Branch name")
+	for _, c := range list.Commits {
+		table.AddRow(c.CommitId, c.CommittedAt, c.CommittedBy, c.BranchName)
+	}
+	_, _ = fmt.Fprint(writer, table)
+	return nil
+}
+
+func constructYamlJson(gen model.GenerationCommits) formattedCommitList {
+	result := formattedCommitList{}
```
Comment (SimonRichardson) [inline cmd/juju/model/listcommits.go]: nit:

```suggestion
	var result formattedCommitList
```

File: cmd/juju/model/showcommit.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+- configuration  made under the branch for each application
+- a summary of how many units are tracking the branch
+
+Examples:
+    juju show-commit 3
+    juju show-commit 3 --utc
+
+See also:
+	list-commits
+    add-branch
```
Comment (SimonRichardson) [inline cmd/juju/model/showcommit.go]: Alignment of the "See also:" list

File: cmd/juju/model/showcommit.go [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    track
+    branch
+    abort
+    diff
+`
+)
+
+//TODO: instead of diffing, i just show the content of config
```
Comment (SimonRichardson) [inline cmd/juju/model/showcommit.go]: Remove the TODO? or put your name on the TODO.

File: cmd/juju/model/showcommit_test.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	err := s.runInit("something")
+	c.Assert(err, gc.ErrorMatches, `encountered problem trying to parse "something" into an int`)
+}
+
+func (s *showCommitsSuite) TestInitMoreArgs(c *gc.C) {
+	args := []string{"1", "2", "3"}
+	err := s.runInit(args...)
+	c.Assert(err, gc.ErrorMatches, "expected exactly 1 commit id, got 3 arguments")
+}
+func (s *showCommitsSuite) getMockValues() coremodel.GenerationCommit {
```
Comment (SimonRichardson) [inline cmd/juju/model/showcommit_test.go]: Can we rename this to something different, like `getGenerationCommitValue`

File: state/modelgeneration.go [hunk 15]
```diff
@@ -703,6 +714,25 @@ func (st *State) Branches() ([]*Generation, error) {
 		return nil, errors.Trace(err)
 	}
 
+	branches := make([]*Generation, len(docs))
+	for i, d := range docs {
+		branches[i] = newGeneration(st, &d)
+	}
+	return branches, nil
+
```
Comment (SimonRichardson) [inline state/modelgeneration.go]: nit: remove line

**Comment by mitechie** [comment]
Thanks for this, really fun to play with. Some feedback

I think you're correct on the "no commits made yet" feedback message. We do this in other places, like if you don't have an active model or controller. In storage it's "No storage to display" so how about sticking with that and doing "No commits to list" (without the headings) when there are none. 

For the list-commits can we sort them desc please as the more recent history at the top is useful. e.g. list-commits | head with the headings vs if you have it at the bottom you've got to let it scroll and lose the headings in the output. I think over time we might make this more interesting and only go back so far without a --all or something but for now I think just sorting reverse would be good. 

For the listing committed at dates, can we do a condensed date format. I'm trying to see if we have a standard for this elsewhere, but I'm tempted to run with what we use in the "juju list-users" output for "Date created" as that's a pretty nice human readable time there. Having the detailed date in show-commit is perfect though. The other way to go would be just a simpler date but then we get into format debates and the like. 

Finally, with us asking folks to play with this I think we need to change the feature flag name to not be "generations" but "branches" to match what the users are interacting with. Any change that's an easy drive-by here?

File: api/modelgeneration/modelgeneration.go [hunk 16]
```diff
@@ -69,6 +69,35 @@ func (c *Client) CommitBranch(branchName string) (int, error) {
 	return result.Result, nil
 }
 
+// ListCommits lists the committed branches commits,
```
Comment (manadart) [inline api/modelgeneration/modelgeneration.go]: "ListCommits returns the details of all committed model branches"

File: api/modelgeneration/modelgeneration.go [hunk 17]
```diff
@@ -69,6 +69,35 @@ func (c *Client) CommitBranch(branchName string) (int, error) {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if result.Error != nil {
+		return nil, errors.Trace(result.Error)
+	}
+	return generationCommitsFromResults(result), nil
+}
+
+// CommitBranch commits the branch with the input name to the model,
```
Comment (manadart) [inline api/modelgeneration/modelgeneration.go]: Copy-pasted.

File: apiserver/facades/client/modelgeneration/interface.go [hunk 18]
```diff
@@ -26,6 +26,8 @@ type Model interface {
 	AddBranch(string, string) error
 	Branch(string) (Generation, error)
 	Branches() ([]Generation, error)
+	CommittedBranch(int) (Generation, error)
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/interface.go]: I would be in favour of calling these `Generation` and `Generations` in the interest of terseness, but that isn't crucial right now.

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 19]
```diff
@@ -276,6 +289,72 @@ func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults
 	return result, nil
 }
 
+// ShowCommit will return details a commit given by its generationId
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: "returns details of the branch with the input generation ID."

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 20]
```diff
@@ -238,7 +250,8 @@ func (api *API) AbortBranch(arg params.BranchArg) (params.ErrorResult, error) {
 // including units on the branch and the configuration disjoint with the
 // master generation.
 // An error is returned if no in-flight branch matching in input is found.
-func (api *API) BranchInfo(args params.BranchInfoArgs) (params.GenerationResults, error) {
+func (api *API) BranchInfo(
+	args params.BranchInfoArgs) (params.GenerationResults, error) {
 	result := params.GenerationResults{}
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: Something I think we *should* do, is rename these:
- `GenerationResults` to `BranchResults`.
- `GenerationCommitResult` to `GenerationResult`.
- `GenerationcommitResults` to `GenerationResults`.

File: core/model/generation.go [hunk 21]
```diff
@@ -57,6 +58,19 @@ type GenerationApplication struct {
 	ConfigChanges map[string]interface{} `yaml:"config"`
 }
 
+// GenerationCommitApplication represents committed changes to an application
+// made under a generation.
+type GenerationCommitApplication struct {
```
Comment (manadart) [inline core/model/generation.go]: Just use `GenerationApplication` for this.
- Add "omitempty" to `UnitProgress`
- `UnitDetail` can hold the same data as `UnitsTracking`. `UnitsPending` will always be empty.

**Comment by nammn** [comment]
!!build!!

**Comment by nammn** [comment]
thanks for the review of this chunk!

$$merge$$

**Pull request merged**